[custom_ops] expose torch.library.register_torch_dispatch

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130261
* #130064

This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

zou3519 committed:
[custom_ops] expose torch.library.register_torch_dispatch

This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] expose torch.library.register_torch_dispatch"


This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented on torch/library.py:
```diff
@@ -870,6 +871,84 @@ def register_autograd(
+        >>>         return func(*args, **kwargs)
+        >>>
+        >>> @torch.library.register_torch_dispatch("mylib::foo", MyMode)
+        >>> def _(mode, func, types, args, kwargs):
```
TODO: add a section to extending.rst and link it from here

zou3519 replied:
done

zou3519 committed:
Update on "[custom_ops] expose torch.library.register_torch_dispatch"


This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on docs/source/notes/extending.rst:
```diff
@@ -914,6 +914,24 @@ It is also possible to add `new` native functions using :mod:`torch.library`. Th
+When a user calls an operator with inputs that have ``__torch_dispatch__``, that call
+may be forwarded to the ``__torch_dispatch__``. args and kwargs get normalized before
+the call to ``__torch_dispatch__``, that is:
+- the ``kwargs`` consist of keyword-only arguments in the operator's schema
```
```suggestion
- the ``kwargs`` consist of keyword-only arguments in the operator's schema, it is only passed if it is not equal to the default value provided in the schema.
```

albanD replied:
Or something similar.

albanD commented on docs/source/notes/extending.rst:
```diff
@@ -914,6 +914,24 @@ It is also possible to add `new` native functions using :mod:`torch.library`. Th
+may be forwarded to the ``__torch_dispatch__``. args and kwargs get normalized before
+the call to ``__torch_dispatch__``, that is:
+- the ``kwargs`` consist of keyword-only arguments in the operator's schema
+- the ``args`` consists of all other arguments, no matter how they were called
```
```suggestion
- the ``args`` consists of all other arguments, no matter how they were passed to the op
```

albanD reviewed (approved):
Sounds good!
Only small nits on the doc

zou3519 committed:
Update on "[custom_ops] expose torch.library.register_torch_dispatch"


This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented:
Ho we also remove positional? lol good to know!

zou3519 commented:
> Ho we also remove positional? lol good to know!

Yes, I'm not a big fan lol

zou3519 committed:
Update on "[custom_ops] expose torch.library.register_torch_dispatch"


This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] expose torch.library.register_torch_dispatch"


This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_ops] expose torch.library.register_torch_dispatch"


This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

izaitsevfb commented:
@pytorchbot revert -m "depends on #130064 which needs to be reverted" -c ghfirst

izaitsevfb commented:
@pytorchbot revert -m "depends on #130064 which needs to be reverted" -c ghfirst

zou3519 committed:
Update on "[custom_ops] expose torch.library.register_torch_dispatch"


This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge -f "unrelated failure"

zou3519 committed:
Update on "[custom_ops] expose torch.library.register_torch_dispatch"


This is the API for defining the interaction between a torch_dispatch
class and a custom op. Taking API bikeshedding.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
